### PR TITLE
Bug 1164058 - Graph chooser should provide a friendlier interface when no graphs specified

### PR DESCRIPTION
**Pull request by MikeLing**
[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/562)

### CODE REVIEW

File: ui/css/perf.css [hunk 1]
```diff
@@ -85,6 +85,37 @@ to { opacity: 0; }
...
+    text-align:center;
+    font-size:25px;
+    padding-bottom:40px;
+    padding-top:15px;
+    
+}
+.panel-content button{
+    text-align:middle;
+    font-size:35px;
+}
```
Comment (tojonmz) [inline ui/css/perf.css]: At least based on current treeherder.css conventions, I believe properties should have a single character whitespace after the `:`. Some of the above do, some don't but I think should. Same with the trailing `{` after the element selectors.

File: ui/partials/perf/graphsctrl.html [hunk 2]
```diff
@@ -23,11 +23,20 @@
...
         <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add more test data</button>
     </div>
-    <div id="data-display">
-      <select ng-model="myTimerange" ng-options="timerange.text for timerange in timeranges track by timerange.value" ng-change="timeRangeChanged()">
+        <div class="panel-center-layout">
+          <div class="panel panel-info" ng-show="seriesList.length == 0">
+            <div class="panel-heading">No graphs date</div>
+            <div class="panel-content">
+            <p >Without any graphs been specified</p>
+            <button type="button" class="btn btn-primary btn-lg btn-block" ng-click="addTestData()"> Add more test </button>
```
Comment (tojonmz) [inline ui/partials/perf/graphsctrl.html]: I haven't tested the branch, but at least looking at the markup it seems these two lines should be indented as they are child elements of the parent `<div>`?

**Comment by MikeLing** [comment]
@tojonmz  Thank you for your advice.  :)
Do you have time now?  Could you give me some advices about the [attachment](https://bug1164058.bugzilla.mozilla.org/attachment.cgi?id=8609817)  in bugzilla?

I think the "add more tests" button is too big but I don't want too much blank lines in this dialog.

**Comment by tojonmz** [comment]
I defer to @wlach he has a good eye :) My take on the [attachment](https://bug1164058.bugzilla.mozilla.org/attachment.cgi?id=8609817) - I am not sure that the light blue div on top and the center white div or their text contents are necessary? Since the blue action button text communicates the same thing.

My take for a default landing experience would be to present an empty graph so the user is cued what their eventual UI dimensions will be, when populated with data. And have our existing [+] button, with suitable wording.

(crude mockup)

crudelandingmock


But please wait until you hear back from @wlach before making changes, as he has a high level view of Perf UI.

File: ui/partials/perf/graphsctrl.html [hunk 3]
```diff
@@ -23,11 +23,20 @@
           </td>
         </tr>
       </table>
-      <button class="btn btn-primary" ng-click="addTestData()">
+      <button class="btn btn-primary" ng-click="addTestData()" ng-hide="seriesList.length == 0">
         <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add more test data</button>
     </div>
-    <div id="data-display">
```
Comment (wlach) [inline ui/partials/perf/graphsctrl.html]: Without the previously-present "data-display" div the flex flow doesn't work correctly, you can see this if you try adding a graph.

**Comment by wlach** [comment]
Ok, on second thought, let's give @tojonmz's approach a shot, maybe with a bit of explanatory text. Perhaps something like this: 

screenshot from 2015-05-25 11 39 54

**Comment by MikeLing** [comment]
When I work as Will told me in last comment, I found maybe a little change can be more friendly and more interesting than just put two blank blocks inside there. So, I create two version interface. Please tell me which one is better, then I will commit the appropriate PR. :)
picture1
picture2

**Comment by wlach** [comment]
Hi Mike, sorry I should have mentioned that it shouldn't be necessary to use any custom CSS to simulate a graph, since we can just plot one with an empty set. We just need to tweak some of the javascript.

We're actually plotting the graph here:

https://github.com/mozilla/treeherder/blob/master/ui/js/graphs.js#L509

Which is called from here when data is specified:

https://github.com/mozilla/treeherder/blob/master/ui/js/graphs.js#L594

Just below that, we cover the case where nothing is specified:

https://github.com/mozilla/treeherder/blob/master/ui/js/graphs.js#L596

We can just make a call to addSeriesList there, something like:

        } else {
          $scope.seriesList = [];
          // add an empty set of series to show graph with no data
          addSeriesList([]);
        }

You'll need to move the zoom handling outside the conditional block, so it always gets set (we should have done this anyway):

        if ($stateParams.zoom) {
            $scope.zoom = JSON.parse($stateParams.zoom);
        } else {
            $scope.zoom = {};
        }

I don't think we need to say "this is empty" inside the graph area. That's pretty obvious from the content on the left, I think. :)

**Comment by tojonmz** [comment]
> I don't think we need to say "this is empty" inside the graph area. That's pretty obvious from the content on the left, I think. :)

I agree with Will. @MikeLing I would also check wlach's mockup in terms of the relative font size of that top left "Nothing here yet." text. I think Will's mockup is using the same font size as the drop down? Font size consistency is always a good thing.

Unrelated - I wonder if it might be handy if the 'Add test data' button lived above the series labels? That way if the user adds lots of series the button doesn't get pushed off screen or move around. My workflow below is probably not representative though. I defer to @wlach :)

lotsofseries

**Comment by wlach** [comment]
Let's leave the button where it is for now. I'd prefer if the button was less prominent than the series labels, and there usually aren't that many graphs. We can always revisit this later. :)

File: ui/js/graphs.js [hunk 4]
```diff
@@ -594,6 +589,12 @@ perf.controller('GraphsCtrl', [
           addSeriesList(partialSeriesList);
         } else {
           $scope.seriesList = [];
+          addSeriesList([]);
+        }
+        if ($stateParams.zoom) {
```
Comment (wlach) [inline ui/js/graphs.js]: This should be above the if clause, not below.

File: ui/partials/perf/graphsctrl.html [hunk 5]
```diff
@@ -24,7 +27,7 @@
         </tr>
       </table>
       <button class="btn btn-primary" ng-click="addTestData()">
-        <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add more test data</button>
+        <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add <span ng-show="seriesList.length > 0">more</span> test data</button>
     </div>
     <div id="data-display">
       <select ng-model="myTimerange" ng-options="timerange.text for timerange in timeranges track by timerange.value" ng-change="timeRangeChanged()">
```
Comment (wlach) [inline ui/partials/perf/graphsctrl.html]: While we're here, let's disable this element if the seriesList is empty. We can use ngDisabled for this:

https://docs.angularjs.org/api/ng/directive/ngDisabled

basically I think we just need to add this to the select element:

ng-disabled="seriesList.length == 0"

File: ui/partials/perf/graphsctrl.html [hunk 6]
```diff
@@ -24,10 +27,10 @@
         </tr>
       </table>
       <button class="btn btn-primary" ng-click="addTestData()">
-        <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add more test data</button>
+        <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Add <span ng-show="seriesList.length > 0">more</span> test data</button>
     </div>
     <div id="data-display">
-      <select ng-model="myTimerange" ng-options="timerange.text for timerange in timeranges track by timerange.value" ng-change="timeRangeChanged()">
+      <select ng-model="myTimerange" ng-options="timerange.text for timerange in timeranges track by timerange.value" ng-change="timeRangeChanged() " ng-disabled="seriesList.length == 0">
```
Comment (wlach) [inline ui/partials/perf/graphsctrl.html]: Looks like you added an extra space after timeRangeChanged() which we don't want --> ng-change="timeRangeChanged() "

File: ui/js/graphs.js [hunk 7]
```diff
@@ -565,7 +565,12 @@ perf.controller('GraphsCtrl', [
     ThOptionCollectionModel.get_map().then(
       function(_optionCollectionMap) {
         optionCollectionMap = _optionCollectionMap;
-
+
```
Comment (wlach) [inline ui/js/graphs.js]: Looks like you unintentionally added some whitespace here.

File: ui/js/graphs.js [hunk 8]
```diff
@@ -565,7 +565,11 @@ perf.controller('GraphsCtrl', [
     ThOptionCollectionModel.get_map().then(
       function(_optionCollectionMap) {
         optionCollectionMap = _optionCollectionMap;
-
```
Comment (wlach) [inline ui/js/graphs.js]: I would put a blank line here (just without any indentation)

File: ui/js/graphs.js [hunk 9]
```diff
@@ -565,7 +565,11 @@ perf.controller('GraphsCtrl', [
...
+        if ($stateParams.zoom) {
+          $scope.zoom = JSON.parse($stateParams.zoom);
+        } else {
+          $scope.zoom = {};
+        }
```
Comment (wlach) [inline ui/js/graphs.js]: Likewise, there shoulkd be a blank line after this.

File: ui/partials/perf/graphsctrl.html [hunk 10]
```diff
@@ -1,6 +1,9 @@
 <div class="container-fluid">
   <div class="ph-horizontal-layout">
     <div id="graph-chooser">
+      <div ng-show="seriesList.length == 0">
+          <p>Nothing here yet</p>
```
Comment (wlach) [inline ui/partials/perf/graphsctrl.html]: This is 2 characters indented too many (compare with rest of html content)

**Comment by wlach** [comment]
This was close enough that I just fixed the last few nits myself and pushed. Please have a look at the PR for more details on what should have been changed though.

**Pull request closed without merging**